service/s3/s3crypto/envelope.go

```diff
@@ -0,0 +1,91 @@
+package s3crypto
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// Envelope encryption starts off by generating a random symmetric key using AES CBC,
+// AES CTR, or AES GCM. We then generate a random IV based off the encryption cipher
+// chosen. The master key that was provided, whether by the user or KMS, will be used
+// to encrypt the randomly generated symmetric key and base64 encode the iv. This will
+// allow for decryption of that same data later.
+type Envelope struct {
+	// IV is the randomly generated IV base64 encoded.
+	IV []byte `json:"x-amz-iv"`
+	// CipherKey is the randomly generated cipher key.
+	CipherKey []byte `json:"x-amz-key"`
+	// MaterialDesc is a description to distinguish from other envelopes.
+	MaterialDesc string `json:"x-amz-matdesc"`
+	Meta         meta   `json:"-"`
+}
+
+type meta struct {
+	Bucket    string
+	Request   *request.Request
+	ObjectKey string
+}
+
+// SaveStrategy is how the data's metadata wants to be saved
+type SaveStrategy interface {
+	Save(Envelope)
+}
+
+// default s3 key
+const instructionKey = ".instruction"
+
+type s3SaveStrategy struct {
+	client                *s3.S3
+	InstructionFileSuffix *string
+}
+
+// NewS3SaveStrategy returns a new strategy that will save to S3
+func NewS3SaveStrategy(p client.ConfigProvider, suffix *string) SaveStrategy {
```

Comment:
May want to consider using functional options instead of pointers here for optional parameters. This allows us to grow the options later own, and have the power to merge optional parameters without dealing with pointers.